webrtc/README.md

```diff
@@ -0,0 +1,412 @@
+
+  No, a browser uses a new UDP port for each `RTCPeerConnection`.
+
+- Can _Browser_ control the lifecycle of its local TLS certificate, i.e. can
+  _Browser_ use the same TLS certificate for multiple WebRTC connections?
+
+  Yes. For the lifetime of the page, one can generate a certificate once and
+  reuse it across connections. See also
+  https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/RTCPeerConnection#using_certificates
+
+- Can two _Browsers_ exchange their SDP packets via a third server node using
+  Circuit Relay v2 and DCUtR? Instead of exchanging the original SDP packets,
+  could they exchange their multiaddr and construct the remote's SDP packet
+  based on it?
+
+## Multiplexing
+
+The WebRTC browser APIs do not support half-closing of streams nor resets of the
+sending part of streams.
+[`RTCDataChannel.close()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel/close)
+flushes the remaining messages and closes the local write and read side. After
+calling `RTCDataChannel.close()` one can no longer read from the channel. This
+lack of functionality is problematic, given that libp2p protocols running on top
+of transport protocols, like WebRTC, expect to be able to half-close or reset a
+stream. See [Connection Establishment in
+libp2p](https://github.com/libp2p/specs/blob/master/connections/README.md#definitions).
+
+To support half-closing and resets of streams, libp2p WebRTC uses message
+framing. Messages on a `RTCDataChannel` are embedded into the Protobuf message
+below and sent on the `RTCDataChannel` prefixed with the message length in
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint spec][uvarint-spec].
+
+It is an adaptation from the [QUIC RFC]. When in doubt on the semantics of
+these messages, consult the [QUIC RFC].
+
+``` proto
+syntax = "proto2";
+
+package webrtc.pb;
+
+message Message {
+  enum Flag {
+    // The sender will no longer send messages on the stream.
+    FIN = 0;
+    // The sender will no longer read messages on the stream. Incoming data is
+    // being discarded on receipt.
+    STOP_SENDING = 1;
+    // The sender abruptly terminates the sending part of the stream. The
+    // receiver can discard any data that it already received on that stream.
+    RESET_STREAM = 2;
+  }
+
+  optional Flag flag=1;
+
+  optional bytes message = 2;
+}
+```
+
+Note that "a STOP_SENDING frame requests that the receiving endpoint send a
+RESET_STREAM frame.". See [QUIC RFC - 3.5 Solicited State
+Transitions](https://www.rfc-editor.org/rfc/rfc9000.html#section-3.5).
+
+Encoded messages including their length prefix MUST NOT exceed 16kiB to support
+all major browsers. See ["Understanding message size
+limits"](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Using_data_channels#understanding_message_size_limits).
+Implementations MAY choose to send smaller messages, e.g. to reduce delays
```

Comment:
Sending messages larger than the MTU introduces HoL blocking, which is exacerbated by the fact that multiple streams might be sending large messages for multiple streams at the same time and applying some kind of round-robin logic.

This suggests that if possible, one should avoid sending anything larger than the MTU. Is there any way to query the browser so we can pack optimally sized packets?

Reply:
> Is there any way to query the browser so we can pack optimally sized packets?

I couldn't find anything MTU related on https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel. I doubt that the browser exposes anything, calculating the payload size might not be straightforward, since the browser would have to take the SCTP framing into account.

This leaves us with 2 bad options:
1. Send `Message`s that are (potentially) larger than the MTU, leading to HoL blocking between streams (even in the absence of packet loss).
2. Send `Message`s that are definitely larger than the MTU, leading to packets being put on the wire that are smaller than the MTU, thereby reducing the bandwidth.

Reply:
https://www.rfc-editor.org/rfc/rfc8831#section-6.6-8

Reply:
https://www.rfc-editor.org/rfc/rfc8831#section-6.6-8 allows SCTP to interleave messages.

Reply:
> https://www.rfc-editor.org/rfc/rfc8831#section-6.6-8 allows SCTP to interleave messages.

Interleaving makes the problem worse, it's what introduces the HoL blocking between stream in the absence of packet loss, as you need to receive the entire Protobuf before you're able to read any of the data.

Reply:
@marten-seemann I am not sure I understand your most recent comment. How does interleaving of packets from different streams make head-of-line-blocking worse? In what sense does it reduce utilization of the underlying connection?

Reply:
Not sure where your concern about the utilization of the underlying connection comes from, that’s not what HoLB is about.

The problem is that you have to receive all ~12 packets that the 16 kB protobuf will be split into, before you can read _any_ of the data you received. If your WebRTC stack now decided to switch to sending on another stream (“interleaving”) after sending 10 of these packets, you’ll have a bunch of data sitting in your stream buffer which has to wait until your WebRTC stack decides to send the other 2 packets for that stream.

Obviously, this problem gets a lot worse once you factor in packet loss (that’s the “classical” HoLB that TCP suffers from). I was just pointing out that in addition to that problem, this proposal introduces HoLB _even in the absence of packet loss_.

Reply:
> The problem is that you have to receive all ~12 packets that the 16 kB protobuf will be split into, before you can read _any_ of the data you received. If your WebRTC stack now decided to switch to sending on another stream (“interleaving”) after sending 10 of these packets, you’ll have a bunch of data sitting in your stream buffer which has to wait until your WebRTC stack decides to send the other 2 packets for that stream.

:+1:

Unfortunately, at this point, I don't see how we could make an informed decision on the best Protobuf size, given that we don't have information like the MTU available.

I suggest we delay this optimization until we have data on real world usage. Implementations may then follow heuristics on how much data they pack into a single framed (protobuf) message. For now, the specification only sets an upper bound, not a lower bound on the Protobuf size.

Today https://github.com/melekes/rust-libp2p/pull/8 follows the simple approach of a new Protobuf message per upper layer write. We may iterate on this in the future, i.e. measure whether the overhead of the message framing is too high.

Reply:
> We may iterate on this in the future, i.e. measure whether the overhead of the message framing is too high.

This is NOT about message framing overhead. This is about HoLB.

Shipping a new transport protocol that exhibits HoLB orders of magnitude worse than TCP (which was one of the reasons people started working on QUIC!) sounds like a non-starter to me. The HoLB is purely the consequence of the choice of an inadequate framing format (Protobuf) for this use case. It’s not a hard problem to solve, and I’ve suggested 3 different approaches how to get rid of the HoLB in https://github.com/libp2p/specs/pull/412#pullrequestreview-1117390305.

Reply:
> > We may iterate on this in the future, i.e. measure whether the overhead of the message framing is too high.
> 
> This is NOT about message framing overhead. This is about HoLB.

I think we are on the same page here. Choosing a smaller message framing size does reduce the potential of head-of-line-blocking at the expense of message framing overhead. Please correct me if I am wrong.

> It’s not a hard problem to solve, and I’ve suggested 3 different approaches how to get rid of the HoLB in https://github.com/libp2p/specs/pull/412#pullrequestreview-1117390305.

Citing the 3 options below for others plus one follow up comment:

> Option 1: One way to avoid this issue would be to introduce a control stream, on which stream resets are sent (at the cost of introducing HoL blocking between stream resets)

> Option 2: or alternatively to have one control stream per data stream (at the cost of doubling the number of streams. Probably not worth it)

Do I understand correctly that one would also communicate `FIN` and `STOP_SENDING` next to `RESET_STREAM` on that control stream? For `FIN` we would need to communicate offsets to tell the remote at what point to not expect more data on the data stream, right?

> Option 3: Obviously, we can also use a serialization format that doesn’t require the entire “thing” to arrive before you can parse it, see for example the design of HTTP/3 DATA frames (as HTTP/3 encounters exactly the same problem).

Reply:
> I think we are on the same page here. Choosing a smaller message framing size does reduce the potential of head-of-line-blocking at the expense of message framing overhead. Please correct me if I am wrong.

That’s correct. Unfortunately, there’s a different problem. If you want to eliminate HoLB, you need to choose sub-MTU packets. I’m not concerned about framing overhead. What I’m concerned about is that there’s no way of learning the MTU, so to be on the safe side, you’d have to send messages around 1100-1200 bytes or so. However, sending sub-MTU packets will decrease the bandwidth you’re likely to achieve, as congestion control (to first order approximation) acts upon packets and not bytes sent (that’s why transports do PLPMTUD).

> Do I understand correctly that one would also communicate FIN and STOP_SENDING next to RESET_STREAM on that control stream? For FIN we would need to communicate offsets to tell the remote at what point to not expect more data on the data stream, right?

Yes, you’d have to communicate all kinds of stream operations, not just RESET_STREAM. Thinking a bit more about it, it gets ugly really quick.

> Citing the 3 options below for others plus one follow up comment:

I think we should go with option 3 :)
I see the appeal of going with the same serialization (Protobuf) everywhere, but I think in this case the price we’d pay for that is too high. Thus shipping our own format sound like the next-best option.

Reply:
I did some research and found this: https://github.com/JitterCompany/jitter_frame_format

The idea is that your data payload is encoded in something like base64. This allows you to pick magic sequences / bytes that are outside of the base64 range which can mark the beginning and end of your frame headers.

This allows you to send packets without knowing their length in advance at the cost of bandwidth and some encoding and decoding in base64.

Reply:
I added ed5c641 and 42e7a20 documenting my latest comment above.

@marten-seemann the specification now recommends a message size of `<= 1252` based on QUICs recommendation. Let me know what you think.

Reply:
I am resolving here. In case you would like to continue this discussion, please unresolve or start a new comment on the markdown file. //CC @marten-seemann as the discussion initiator.